### PR TITLE
Search: migrate search widget to package

### PR DESCRIPTION
**Pull request by kangzj**
#### Changes proposed in this Pull Request:
The PR moves Jetpack Search widget to the search package. Changes include:

- Moved assets to the search package.
- Text domain changed to `jetpack-search-pkg`
- Moved initialization to `Jetpack_Initializer`
- Some other changes regarding namespace change

PHP linter tried to arrange the order of the functions which lead to possibly hard review, and had to re-order and commit with `--no-verify`.

#### Jetpack product discussion
pcNPJE-oI-p2

#### Does this pull request change what data or activity we track or use?
no

#### Testing instructions:
- Navigate to a site with Jetpack Search subscription
- Disable Instant search
- Add Jetpack Search widget to sidebar
- Ensure assets load on the frontend: `search-widget-frontend.js` and `search-widget-frontend.css` (the style file is concatenated in /wp-content/plugins/jetpack/css/jetpack.css)
- Ensure Jetpack Search widget works in widget editor and customizer, meaning you could add/remove filters and save changes
- Ensure Jetpack Search widget works on the frontend for both Instant and Classic Search

Note: please use a theme that supports  widgets (e.g. theme twentytwenty) as FSE themes don't have widget areas.

### CODE REVIEW

File: projects/packages/search/src/dashboard/class-dashboard.php [hunk 1]
```diff
@@ -47,7 +47,8 @@ class Dashboard {
 	 * @param Automattic\Jetpack\Search\Module_Control $module_control - Module_Control instance.
 	 */
 	public function __construct( $plan = null, $connection_manager = null, $module_control = null ) {
-		$this->plan               = $plan ? $plan : new Plan();
+		$this->plan = $plan ? $plan : new Plan();
+		// TODO: 'jetpack-search' better to be the current plugin where the package is running.
 		$this->connection_manager = $connection_manager ? $connection_manager : new Connection_Manager( 'jetpack-search' );
```
Comment (jsnmoon) [inline projects/packages/search/src/dashboard/class-dashboard.php:52]: I'm not quite understanding your comment here.

Do you mean we'd want to change `'jetpack-search'` to reflect which plugin we're running in? (E.g. `'jetpack-search'` in the standalone plugin, `'jetpack'` in the mainline plugin?)

File: projects/packages/search/src/initializers/class-jetpack-initializer.php [hunk 2]
```diff
@@ -77,4 +80,21 @@ public static function is_connected() {
...
+
+	/**
+	 * Register the widget if Jetpack Search is available and enabled.
+	 */
+	public static function jetpack_search_widget_init() {
+		if (
+		! self::is_connected()
+		|| ! self::is_search_supported()
+		|| ! ( new Module_Control() )->is_active()
+		) {
```
Comment (jsnmoon) [inline projects/packages/search/src/initializers/class-jetpack-initializer.php:92]:
```suggestion
		if (
			! self::is_connected()
			|| ! self::is_search_supported()
			|| ! ( new Module_Control() )->is_active()
		) {
```

File: projects/packages/search/src/initializers/class-jetpack-initializer.php [hunk 3]
```diff
@@ -77,4 +80,21 @@ public static function is_connected() {
...
+			return;
+		}
+
+		// There won't be multiple widgets registered when Search stand alone plugin registers it again.
+		// Because the function tests the hash of the class, if they are the same, just register again.
```
Comment (jsnmoon) [inline projects/packages/search/src/initializers/class-jetpack-initializer.php:97]: Good to know!

File: projects/packages/search/src/widgets/class-search-widget.php [hunk 4]
```diff
@@ -131,28 +117,35 @@ public function activate_search() {
 	public function widget_admin_setup() {
 		wp_enqueue_style(
 			'widget-jetpack-search-filters',
-			plugins_url( 'search/css/search-widget-admin-ui.css', __FILE__ ),
+			plugins_url( 'css/search-widget-admin-ui.css', __FILE__ ),
 			array(),
-			JETPACK__VERSION
+			// TODO: We'll need our own version number.
```
Comment (jsnmoon) [inline projects/packages/search/src/widgets/class-search-widget.php:122]: Could you explain what you mean here? Are you referring to making `Helper::get_asset_version` independent of the Jetpack plugin?

**Review by jsnmoon — COMMENTED** [review]
Works as described; left some comments in-line with the code.

I also discovered a curious bug: Switching from Instant Search to Classic Search leaves the Jetpack Search widget in a broken state, omitting the search box and sort dropdown when rendered. This isn't a regression caused by this PR but wanted to mention it just in case.

> Screen Shot 2022-01-19 at 6 35 53 PM
> Screen Shot 2022-01-19 at 6 36 13 PM

EDIT: Opened https://github.com/Automattic/jetpack/issues/22442.

File: projects/packages/search/src/widgets/class-search-widget.php (hunk 4 above)
Comment (kangzj) [reply projects/packages/search/src/widgets/class-search-widget.php:122]: Yep that's right - I figure I could just change it to use the `Asset` class, and `Helper::get_asset_version` is to be removed in #22419.

File: projects/packages/search/src/dashboard/class-dashboard.php (hunk 1 above)
Comment (kangzj) [reply projects/packages/search/src/dashboard/class-dashboard.php:52]: Yes, you are right - sorry for the poor expression. I guess we'll need a flag or something to mark which env it is in.

File: projects/packages/search/src/widgets/class-search-widget.php (hunk 4 above)
Comment (jsnmoon) [reply projects/packages/search/src/widgets/class-search-widget.php:122]: Excellent!

File: projects/packages/search/src/dashboard/class-dashboard.php (hunk 1 above)
Comment (jsnmoon) [reply projects/packages/search/src/dashboard/class-dashboard.php:52]: That makes sense, thanks for the clarification 🙂

**Review by jsnmoon — COMMENTED** [review]
I also noticed that the filters weren't initially populating for the Search widget with Classic Search.

<table>
<tr>
<td>Classic Search at `/`
<td>Classic Search at `/?s=`
<td>Instant Search
</tr>
<tr>
<td>Screen Shot 2022-01-20 at 4 40 39 PM
Note the missing filters.
<td>Screen Shot 2022-01-20 at 4 37 30 PM
Populated filters.
<td>Screen Shot 2022-01-20 at 4 37 30 PM
Populated filters.
</tr>
</table>

I don't think this is a regression caused by this PR, however.

EDIT: Opened https://github.com/Automattic/jetpack/issues/22441.

**Review by jsnmoon — DISMISSED** [review]
Great work! Changes are reasonable and test well on my local docker setup as well as on a Jurassic Ninja instance.

File: projects/packages/search/src/widgets/class-search-widget.php [hunk 5]
```diff
@@ -182,43 +175,43 @@ public function enqueue_frontend_scripts() {
...
-			JETPACK__VERSION,
+			Helper::get_asset_version( __FILE__ ),
 			true
 		);
 
+		// TODO: No need to enqueue the style for Jetpack the plugin as it's concatenated together.
+		// @see https://github.com/Automattic/jetpack/blob/b3de78dce3d88b0d9b283282a5b04515245c8057/projects/plugins/jetpack/tools/builder/frontend-css.js#L52.
 		wp_enqueue_style(
 			'jetpack-search-widget',
 			plugins_url( 'search/css/search-widget-frontend.css', __FILE__ ),
```
Comment (jeherve) [inline projects/packages/search/src/widgets/class-search-widget.php:186]: Should we update the paths here as well, and remove `search`? I think that part of the path doesn't exist anymore.

**Review by jeherve — CHANGES_REQUESTED** [review]

File: projects/packages/search/src/widgets/class-search-widget.php (hunk 5 above)
Comment (kangzj) [reply projects/packages/search/src/widgets/class-search-widget.php]: Thanks for pointing that out @jeherve 
Changed to use `Assets` to enqueue assets.

**Review by jeherve — COMMENTED** [review]
I'm not sure if I'm doing something wrong, but I don't see the widget's JS and CSS being enqueued on the frontend, whether concatenation is enabled or disabled (with `define( 'SCRIPT_DEBUG', true );`). Does it work for you?

I also get to see this warning on a JN site where concatenation is disabled:
```
PHP Warning:  filemtime(): stat failed for /wp-content/plugins/jetpack-dev/jetpack_vendor/automattic/jetpack-search/src/widgets/js/search-widget-admin.js in /wp-content/plugins/jetpack-dev/jetpack_vendor/automattic/jetpack-assets/src/class-assets.php on line 374
```

**Comment by kangzj** [comment]
Hi @jeherve thanks for your feedback. The assets only load for Classic Search as it did. I tried to keep everything as they were. Sorry the instructions weren't so clear. Just updated, would you mind testing it again? Thanks.

**Comment by kangzj** [comment]
> PHP Warning:  filemtime(): stat failed for /wp-content/plugins/jetpack-dev/jetpack_vendor/automattic/jetpack-search/src/widgets/js/search-widget-admin.js in /wp-content/plugins/jetpack-dev/jetpack_vendor/automattic/jetpack-assets/src/class-assets.php on line 374

**Review by jeherve — APPROVED** [review]
Looking good, let's merge this! 🚢

**Comment by anomiex** [comment]
> This possibly is a @anomiex question

The jitm one probably means you didn't build packages/jitm in your local dev environment, so the built file isn't there to be served.

The search one is the opposite, files in packages/search's `src/` are excluded from the mirror so they're not there to be served.

**Comment by kangzj** [comment]
I tested the package on WPCOM which works fine. The changes regarding the search widget will be applied manually to WPCOM after next release. And fusion will be removed accordingly.

cc @jsnmoon @bluefuton

**Pull request merged**